fhir-server/src/main/java/au/csiro/pathling/fhirpath/literal/QuantityLiteralPath.java

```diff
@@ -6,92 +6,225 @@
+      @Nonnull final FhirPath context, @Nonnull final UcumService ucumService) {
+    final Matcher matcher = UCUM_PATTERN.matcher(fhirPath);
     if (!matcher.matches()) {
-      throw new IllegalArgumentException("Quantity literal has invalid format: " + fhirPath);
+      throw new IllegalArgumentException("UCUM Quantity literal has invalid format: " + fhirPath);
     }
-
+    final String fullPath = matcher.group(0);
+    final String value = matcher.group(1);
     final String rawUnit = matcher.group(2);
-    final String unit = StringLiteralPath.fromString(rawUnit, context).getLiteralValue()
+    final String unit = StringLiteralPath.fromString(rawUnit, context).getValue()
         .getValueAsString();
+
+    @Nullable final String validationResult = ucumService.validate(unit);
+    if (validationResult != null) {
+      throw new InvalidUserInputError(
+          "Invalid UCUM unit provided within Quantity literal (" + fullPath + "): "
+              + validationResult);
+    }
+
+    final BigDecimal decimalValue = getQuantityValue(value, context);
+    @Nullable final String display = ucumService.getCommonDisplay(unit);
+
+    return buildLiteralPath(decimalValue, unit, Optional.ofNullable(display), context, fhirPath);
+  }
+
+  @Nonnull
+  public static QuantityLiteralPath fromCalendarDurationString(@Nonnull final String fhirPath,
+      @Nonnull final FhirPath context) {
+    final Matcher matcher = CALENDAR_DURATION_PATTERN.matcher(fhirPath);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException(
+          "Calendar duration literal has invalid format: " + fhirPath);
+    }
+    final String value = matcher.group(1);
+    final String keyword = matcher.group(2);
+
     final Quantity quantity = new Quantity();
-    quantity.setUnit(unit);
+    quantity.setValue(new BigDecimal(value));
+    quantity.setSystem(FHIRPATH_CALENDAR_DURATION_URI);
+    quantity.setCode(keyword);
+
+    return new QuantityLiteralPath(context.getDataset(), context.getIdColumn(), quantity, fhirPath);
+  }
 
+  private static BigDecimal getQuantityValue(final String value, final @Nonnull FhirPath context) {
+    final BigDecimal decimalValue;
     try {
-      final long value = IntegerLiteralPath.fromString(fhirPath, context)
-          .getLiteralValue().getValue();
-      quantity.setValue(value);
+      decimalValue = DecimalLiteralPath.fromString(value, context).getValue().getValue();
     } catch (final NumberFormatException e) {
-      try {
-        final BigDecimal value = DecimalLiteralPath.fromString(fhirPath, context)
-            .getLiteralValue().getValue();
-        quantity.setValue(value);
-      } catch (final NumberFormatException ex) {
-        throw new IllegalArgumentException("Quantity literal has invalid format: " + fhirPath);
-      }
+      throw new IllegalArgumentException("Quantity literal has invalid value: " + value);
     }
+    return decimalValue;
+  }
 
```

Comment:
I think both of the parsing methods above could be refactored out to `QuantityLiteral` in the same way we have `CodingLiteral`

Reply:
Parsing of calendar duration refactored to CalendarDuationUtils. Ucum quantities parsing can be refactored later is need for python interface (see CodingLiteral).